apps/emqx/src/emqx_channel.erl

```diff
@@ -1434,15 +1434,24 @@ terminate({shutdown, kicked}, Channel) ->
     run_terminate_hook(kicked, Channel);
 terminate({shutdown, Reason}, Channel) when
     Reason =:= discarded;
-    Reason =:= takenover;
-    Reason =:= not_authorized
+    Reason =:= takenover
 ->
     run_terminate_hook(Reason, Channel);
 terminate(Reason, Channel = #channel{will_msg = WillMsg}) ->
-    (WillMsg =/= undefined) andalso publish_will_msg(WillMsg),
+    should_publish_will_message(Reason, Channel) andalso publish_will_msg(WillMsg),
     (Reason =:= expired) andalso persist_if_session(Channel),
     run_terminate_hook(Reason, Channel).
 
+should_publish_will_message(TerminateReason, Channel) ->
+    not lists:member(TerminateReason, [
+        {shutdown, kicked},
+        {shutdown, discarded},
+        {shutdown, takeovered},
```

Comment:
```suggestion
        {shutdown, takenover},
```

Reply:
nit: the `takenover` and `discarded`  will be matched  out by the previous clause (Line 1436), so it is useless here?🤔

Reply:
If we filter out  the `not_authorized` here, the problem mentioned here https://github.com/emqx/emqx/pull/8894#discussion_r964334474 is remaining existed?

Reply:
> nit: the takenover and discarded will be matched out by the previous clause (Line 1436), so it is useless here?

Yes, since at this moment we check for `takenover` in the clause just before this one, we won't fall into this case here. 

Keeping it here makes the rule/intention more clear, I think, since that's also a case when we don't want to publish any last will testament messages.  What do you think?

> If we filter out the not_authorized here, the problem mentioned here #8894 (comment) is remaining existed?

Since there's a check below for the connection state not being `idle` nor `connecting`, wouldn't those cover the issue pointed out?

```erlang
not lists:member(info(conn_state, Channel), [idle, connecting])
```

Would we need more information to make that judgement?

Reply:
> Keeping it here makes the rule/intention more clear

:+1:

> Since there's a check below for the connection state not being idle nor connecting, wouldn't those cover the issue pointed out?

It's no help, I think. Because it can't pass the `not lists:members(TerminateReason, [{shutdown, not_authorized}, ...])` check :(

> Would we need more information to make that judgement?

I think the check should be enough now, it's much clearer than before :], but we still need to add some cases to cover them.

Reply:
It makes total sense.  I've added a couple more tests here to check the LWT when `deny_action = disconnect`: https://github.com/emqx/emqx/pull/8930/commits/b7191ca7e380a855b674d035169b2d07a6071b97

The LWT seems to be correctly published in those cases :smiley_cat:

Reply:
I saw that it did work. Sorry, looks like I misled you, because, after the authz check failed, will_msg was sent in `emqx_channel:handle_info({sock_closed..`  and not here (`emqx_channel:terminated/2`) :(

This seems a bit difficult to understand, if not in contact with the implementation of other places.
```erlang
    not lists:member(TerminateReason, [
        {shutdown, kicked},    %% has sent in `handle_call(kicked` first
        {shutdown, discarded}, %% no need to send
        {shutdown, takenover}, %% no need to send
        {shutdown, not_authorized}  %% it is hard to understand why we should not send lwt in this error reason
    ]) andalso
```
Would it be easier to understand to remove it, i.e:?
```erlang
should_publish_will_message(TerminateReason, Channel) ->
    %% the will_msg has not been sent yet
    info(will_msg, Channel) =/= undefiend andalso
    %% must be in authentication succeed state 
        not lists:member(info(conn_state, Channel), [idle, connecting]).
```

the reason for this change is to refer to:
> "The Server SHOULD publish Will Messages promptly after the Network Connection is closed and the Will Delay Interval has passed, or when the Session ends, whichever occurs first. "
> 
> quoted from [mqtt-5.0#3.1.2.5](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901040)

Anyway, this is just some not-so-important thoughts, hope it will be helpful. Sorry again for taking so much time to understand why `not_authorized` works here and thanks for your patience :).

Reply:
nit: I made a commit. IMO, it might be a little easier to understand. https://github.com/HJianBo/emqx/commit/80ddbefe564ef1ff114e1dee521e496e1f4178f6

Reply:
Strangely, for some reason the tests are indicating that functionally it's working :joy_cat: 

But your patch seems better indeed.  It's actually closer to what 4.x does, which was not affected by the original bug in the first place.

Reply:
I've incorporated your patch here: https://github.com/emqx/emqx/pull/8930/commits/db9d654b9dfdd4ef79ecff51b5e8b03fc5946fc2

Thanks!  :beers: